### ISSUE TITLE
[Question] ChangeTracking.Internal tests: what do these abbreviations mean?

### ISSUE THREAD

**Issue by jzabroski** [dotnet/efcore#15529]
# People Tags
* @rowanmiller - I realize you no longer work on EF but my understanding is you were the Program Manager with "EF Classic" rolled out code-first migrations.
* @ajcvickers - I see your hands on most of the source code files I have perused the last couple of hours trying to wrap my head around this.  I also messaged you on twitter.

# Question
I have looked at the following tests in `Microsoft.EntityFrameworkCore.ChangeTracking.Internal` namespace:

- FixupCompositeTest.cs
- FixupTest.cs
- ShadowFixupTest.cs
- ShadowFxFixupTest.cs

Additionally, the `Microsoft.EntityFrameworkCore` namespace has a general specification test containing many of the same method names:

- StoreGeneratedFixupTestBase.cs

Some things that might be helpful for me to contribute:
- What does the phrase "fixup" mean?
- What does the phrase "uni" mean?  Unidirectional?
- What does the abbreviation `DN`  `ParentDN` mean?
- What does the abbreviation `NN` in `ParentNN` mean?
- What does the abbrevation `DN` in `ChildDN` mean?
- What does the abbreviation `NN` in `ChildNN` mean?
- What does the abbreviation `princ` mean?
- What does the abbreviation `uni` mean?
- What does set_principal_nav_set? (Is the word set used in two different ways in `Add_dependent_then_principal_one_to_one_prin_uni_FK_set_principal_nav_set`?).
- etc...
- If I write a .NET Reflection program that looks for all test fixtures with the phrase "Fixup" in it, and output the matrix of classes with identical methods (as a matrix/excel spreadsheet), what would it look like?

# Thoughts on Improving Code Coverage as EFCore Feature Set Expands
Controversial idea: EFCore development has slowed and bug reports have increased exponentially with each release.  This is not due to poor coding effort by Microsoft or open source community, but due to the fact the test cases are not directly expressed combinatoric features.  There is a general testing methodology for scaling code coverage with features called Orthogonal Array Testing Strategy which can potentially scrap some of the boilerplate witnessed above, and potentially fuzz out some bugs/feature gaps in the product like ones I think my team found this week.

I tend to be a systematic thinker, and am curious how open the EF team is to refactoring these tests so that they can be independently extensible as new model mapping features are added.

Moreover, I find it curious that the tests seem to be focused just on change tracking.  I would like tests that test:
1. properties are persisted
2. cascade works as expected
3. queries return the correct result

Thinking out-loud, I wonder if most if not all object models can be compactly expressed as a W-grammar / parsing expression grammar, and we can automatically generate all test combinations from a few sentences in such a grammar.

e.g. something like - this is totally half-baked as I'm in the office at 9:30PM and my wife waiting for me to come home, so please don't Judge me yet.

```
<parent_model_type> :: TPH | TPT
<mapping_strategy> :: Shadow | Foreign Key | OwnedEntity 
<child_model_type> :: <parent_model_type> with <mapping_strategy>
```

**Comment by jzabroski** [comment]
@ajcvickers  I spent all day today trying to figure out what is and isn't supported by Table-Per-Hierarchy.  I've come up with 6 tests - 4 fail, 2 pass but I suspect they are silent failures.

Please see: https://github.com/jzabroski/EntityFrameworkMappingTablePerHierarchy/

**Comment by jzabroski** [comment]
In case I stop caring to update this, and some poor soul tries to implement table-per-hierarchy in the way I describe in my demo:

- It appears you cannot have the same Discriminator name in two separate hierarchies.

**Comment by ajcvickers** [comment]
@jzabroski 
> What does the phrase "fixup" mean?

It means setting navigation properties and/or FKs such that they are in a consistent state.

> What does the phrase "uni" mean? Unidirectional?

Yes, unidirectional.

> What does the abbreviation DN ParentDN mean?

Dependent navigation

> What does the abbreviation NN in ParentNN mean?

No navigations

> What does the abbrevation DN in ChildDN mean?

Dependent navigation

> What does the abbreviation NN in ChildNN mean?

No navigations

> What does the abbreviation princ mean?

Principal

> What does the abbreviation uni mean?

Unidirectional

> What does set_principal_nav_set? (Is the word set used in two different ways in Add_dependent_then_principal_one_to_one_prin_uni_FK_set_principal_nav_set?).
etc...

This means, "add the dependent entity first, then the principal, for a 1:1 relationship with a unidirectional principal navigation property when the FK has already been set and the principal navigation property is also already set."

> If I write a .NET Reflection program that looks for all test fixtures with the phrase "Fixup" in it, and output the matrix of classes with identical methods (as a matrix/excel spreadsheet), what would it look like?

No idea. Did you try it?

----

With regard to test philosophy, we come from a background of using massive matrices of generated tests. We don't do that anymore because it tends to obscure what is really happening in such a way that:
* It can be hard to tell if the test is actually testing the correct things and checking for the correct results
* It's easy to miss things that seem to make sense when considered as part of the matrix, but don't make sense when writing actual application code. One of the most important reason to write tests is to be able to say as a developer, "I know we said it should work this way, but this is a horrible experience, so we really should think again." But big automated matrixes obscure this kind of analysis.
* In our experience, these kinds of tests are hard to debug.
* They can take a very long time to run.

Certainly there are pros and cons of both approaches, but the primary thing we strive for in our tests is that if you read the test code, then you can figure out just from reading the test code what is happening. I'm not saying we always achieve this. Testing is hard. Maintaining tests is hard.

----

With regard to issues you have found, it would be great if you could file GitGub issues for these, complete with a _short_ description, and a small runnable project/solution, code listing, or even some test that we can drop in somewhere such that we can easily reproduce the behavior you are seeing.

**Comment by jzabroski** [comment]
Haha, understood on the _short_ description requirement.  Sorry for so much back and forth recently.  Just like anyone, on a deadline trying to figure out how to elegantly implement something as painlessly and fast as possible.

tl;dr: Will do what you asked.

Long version:

My linked to repository is already written using XUnit.  I had originally wrote them as console programs but it was not much effort to convert to XUnit.  I'm about to update them by converting a few from Facts to Theories to better understand some nuances.  For example, it also seems `HasColumnType` is not compatible with all versions of `HasConversion`.

> Did you try [building a matrix of tests]?

Yes.  But it's worthless at the moment because I don't fully understand all the moving parts/features and how they should/can interact.  For example, in my testing, calling `EnsureDeleted` in the Arrange phase will almost certainly cause the test to fail before the sut (change tracking) is Act'ed upon.  I don't understand why this happens.

> Certainly there are pros and cons of both approaches.

I tend to do what I call the _Wiggle-Waggle Principle_, and I don't see many other engineers who practice similar philosophy because they tend to just use existing toolchains rather than build a new toolchain custom fit for a task.  The Wiggle-Waggle Principle states that I need to be able to wiggle something in the front end, and see what waggles in the backend.  Similarly, I want to waggle something in the backend, and see what wiggles in the frontend.  I suppose the M compiler @AndriySvyryd  worked on for Oslo is similar in that regard, although I never played with it enough to know for sure.  However, the more basic framework I am familiar with is `ASN.1` notation for representing patterns in a Rete network - this is effectively how pattern matching is memoized/works in heavily optimized stream-based complex event processing like StreamBase.

> In our experience, these kinds of tests are hard to debug.

Not sure if helpful/illuminating, but in my experience that was exactly the case, until I realized my problem was an artificial _closed world assumption_.  There are only two solutions to this problem: increase the size of the closed world (effectively download customer data and use it to drive fuzz testing), and creating a snapshot mechanism (similar to Visual Studio Time Travel Debugging) to record and replay issues.
